feat(instancepoller): wire up the instancepoller with the network domain to update provider config

gfouillet opened the pull request:
Wire up the instancepoller with the network domain to update provider config.

This is fairly straight forward. After this PR, the instancepoller will double write incoming information for devices and addresses. 

I also removed the stub tests in instancepoller_test.go because they are covered by state tests. Added some edge cases in state tests to actually cover it, and fix a bug (origin was correctly setted down to machine when no provider id was received, but not setted up to provider when a provider ID was obtained. It is fixed know.



## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

bootstrap a controller, add a model, add machine, deploy application.

Check in both mongodb and dqlite that link layer information and addresses are consistent. 


## Links

**Jira card:** JUJU-7770

hmlanigan reviewed (commented):
questions while the bigger one from Joe is resolved.

gfouillet commented on apiserver/facades/controller/instancepoller/instancepoller_test.go:
```diff
@@ -131,18 +132,6 @@ func (s *InstancePollerSuite) SetUpTest(c *tc.C) {
-- An unseen provider-sourced address should change its origin to "machine".
-- Incoming network config with 2 different devices having the same provider ID should error.
-- A bridge and an ethernet device, both with the same MAC address;
-  SetProviderNetworkConfig is called with a device match the MAC address (no name);
```
In the state layer:

https://github.com/juju/juju/blob/dcc6d9727f4c5e10227f8c5fcd3579f174c6ce7b/domain/network/state/linklayer_merge_test.go#L133-L180

gfouillet commented on domain/network/state/linklayer_merge.go:
```diff
@@ -174,7 +175,16 @@ VALUES ($insert.provider_id, $insert.address_uuid)
+	type uuids []string
+	updateOriginStmt, err := st.Prepare(`
+UPDATE ip_address 
+SET origin_id = 1 -- set origin to provider
```
This method just setup a provider ID by inserting a row in `provider_ip_address`, so this update is locally valid (if something went wrong, the transaction will be rollbacked).

For your whole question, is the whole point of MergeLinkLayerDevices to ensure that this specific method is called for the right addresses.

The instancepoller calls SetProviderConfig with whatever it provides, and the state layer check if there is provider ids and do the whole reconciliation with existig devices.

gfouillet commented:
> Something is not working correctly. In Mongo, we can see that the original machine-sourced addresses get a provider origin when the instance poller runs:
> 
> (...)

Gotcha. I messed up with the incoming data lay out (providerAddressID was passed with the device infos not the addresses, it is why I missed them. Also it has been a bug in the state layer.

I fixed both issue and updated state layer tests to trigger the bug if not fixed.

Let me rebase and ask for a second review. thanks.

hmlanigan commented on apiserver/facades/controller/instancepoller/instancepoller.go:
```diff
@@ -377,3 +388,70 @@ func (a *InstancePollerAPI) AreManuallyProvisioned(ctx context.Context, args par
+func newNetInterface(device network.InterfaceInfo) domainnetwork.NetInterface {
+	return domainnetwork.NetInterface{
+		Name:             device.InterfaceName,
+		MTU:              ptr(int64(device.MTU)),
+		MACAddress:       ptr(device.MACAddress),
+		ProviderID:       ptr(device.ProviderId),
```
question: are we guaranteed the values converted to pointers actually exist?  For `device.GatewayAddress.Value` too.

gfouillet replied:
We don't need this garantee. `ptr` will return a nil ptr if the values is equal to the zero-value of its type.

manadart reviewed (approved):
Seems good now; thanks.
